src/PollingBlockTracker.ts

```diff
@@ -106,9 +120,33 @@ export class PollingBlockTracker
       return this._currentBlock;
     }
     // wait for a new latest block
-    const latestBlock: string = await new Promise((resolve) =>
-      this.once('latest', resolve),
-    );
+    const latestBlock: string = await new Promise((resolve, reject) => {
+      // eslint-disable-next-line prefer-const
+      let onLatestBlockUnavailable: InternalListener;
+      const onLatestBlockAvailable = (value: string | PromiseLike<string>) => {
+        this.#removeInternalListener(onLatestBlockAvailable);
+        this.#removeInternalListener(onLatestBlockUnavailable);
+        this.removeListener('error', onLatestBlockUnavailable);
+        resolve(value);
+      };
+      onLatestBlockUnavailable = () => {
+        // if the block tracker is no longer running, reject
+        // and remove the listeners
+        if (!this._isRunning) {
+          this.#removeInternalListener(onLatestBlockAvailable);
+          this.#removeInternalListener(onLatestBlockUnavailable);
+          this.removeListener('latest', onLatestBlockAvailable);
+          this.removeListener('error', onLatestBlockUnavailable);
+          reject(
+            new Error('Block tracker ended before latest block was available'),
+          );
+        }
+      };
+      this.#addInternalListener(onLatestBlockAvailable);
+      this.#addInternalListener(onLatestBlockUnavailable);
+      this.once('latest', onLatestBlockAvailable);
+      this.on('error', onLatestBlockUnavailable);
```

Comment:
Hmm, I'm thinking that this would not cover the scenario where the block tracker is destroyed before fetching the first block and without throwing errors.

We can probably listen to `_ended` instead

Reply:
It looks like `_ended` could present other issues, because of the weird order of events received:
1. `_start` is emitted
2. Latest block is fetched from the provider
3. `_ended` event listener is executed
4. `latest` event listener is executed

Reply:
Right, so the problem here, as I understand it, is that because `getLatestBlock` now relies on `latest` or `error` to fire in order to resolve the promise it creates, if that never happens because the `fetch` never occurs (or, alternatively, does occur but the promise it returns hasn't resolved yet) then `getLatestBlock` will just hang, even if `destroy` is called?

Reply:
I am slow today :/ It seems that Mark's solution in #286 may solve that case because it grants `destroy` access to the promise that `getLatestBlock` creates, and we can just force-reject it if it hasn't resolved yet.